BloomFilter.md

```diff
@@ -47,23 +47,33 @@ pushdown even in columns with high cardinality and when space is at a premium.
 The initial Bloom filter algorithm in Parquet is implemented using a combination of two
 Bloom filter techniques.
 
-First, the block Bloom filter algorithm from Putze et al.'s "Cache-, Hash- and
-Space-Efficient Bloom filters"[2] is used. This divides a filter into many tiny Bloom
-filters, each one of which is called a "block". In Parquet's initial implementation, each
-block is 256 bits. When inserting or finding a value, part of the hash of that value is
-used to index into the array of blocks and pick a single one. This single block is then
-used for the remaining part of the operation.
-
-Second, within each block, this implementation uses the folklore split Bloom filter
-technique, as described in section 2.1 of "Network Applications of Bloom Filters: A
-Survey"[5]. This divides the 256 bits in each block up into eight contiguous 32-bit lanes
-and sets or checks one bit in each lane.
+First, the block Bloom filter algorithm from Putze et al.'s [Cache-, Hash- and Space-Efficient
+Bloom filters](http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf) is used.
+The block Bloom filter consists of a sequence of small Bloom filters, each of which can fit
+into one cache-line. For best performance, those small Bloom filters are loaded into memory
+cache-line-aligned. For each potential element, the first hash value selects the Bloom filter block
+to be used. Additional hash values are then used to set or test bits as usual, but only inside
+this one block. As for Parquet's initial implementation, each block is 256 bits. When inserting or
+finding value, the first hash of that value is used to index into the array of blocks and pick a
+single one. This single block is then used for the remaining part of the operation.
+
+Second, the remaining part of the operation within the single block uses the folklore split Bloom
+filter technique, as described in section 2.1 of [Network Applications of Bloom Filters:
+A Survey](https://www.eecs.harvard.edu/~michaelm/postscripts/im2005b.pdf). Instead of having one
+array of size `m` shared among all the hash functions, each hash function has a range of `m/k`
+consecutive bit locations disjoint from all the others. The total number of bits is still
+<em>m</em>, but the bits are divided equally among the `k` hash functions. This approach
+can be useful for implementation reasons, for example, dividing the bits among the hash functions
+may make parallelization of array accesses easier and take utilization of SSE. As for Parquet's
+implementation, it divides the 256 bits in each block up into eight contiguous 32-bit lanes and
+sets or checks one bit in each lane.
 
 #### Algorithm
 In the initial algorithm, the most significant 32 bits from the hash value are used as the
 index to select a block from bitset. The lower 32 bits of the hash value, along with eight
 constant salt values, are used to compute the bit to set in each lane of the block. The
```

Comment:
What is the purpose of salting here?

Reply:
As the paper notes, the multiply-shift hashing technique is highly dependent on the fixed multiplicand. This follows the common technique of calling a non-key parameter to a hash function the "salt" or the "seed".

Reply:
A short summary of the multiply-shift hashing technique would be welcome here as the linked paper is 33 pages long.

Reply:
OK, I replaced it with wiki https://en.wikipedia.org/wiki/Hash_function#Multiplicative_hashing.

Reply:
I'm sorry, but from the current specification it is still not clear to me what the salt is used for. I have a theory based on @jbapple's comment (see below), please confirm or correct it, but either way it should be described in the document as well.

So here is my theory: A bloom filter requires `k` different hash functions, each mapping the same input value `x` to different output values. Instead of `k` different hash functions, we can employ a single bivariate function using the same `x` as one of the input values but `k` different values for the other input variable. These `k` different value are the salt. So, am I even close to the correct answer? :)

Reply:
Basically, you are in the right direction. Think it another way, different salts are used to construct different hash functions, as the form in the referred wiki

hash<sub>i</sub>(x) = salt<sub>i</sub> * x >> y

Since the target hash value is `[0, 31]`, so we right shift `y = 27` bits. As a result, we get eight hash values, which are indexes of the tiny bloom filter.

Reply:
Makes sense, thanks. I consider an indexed function with a single parameter the same as non-indexed one with two parameters, so we are on the same page. Could you please incorporate this explanation into the spec? Additionally, it may be worth mentioning that in the linked wikipedia article, `a` corresponds to the salt, since the article does not use the same terminology. Thanks!